src/config.js

```diff
@@ -276,9 +276,41 @@ export default class Config {
       networkConcurrency: this.networkConcurrency,
       networkTimeout: this.networkTimeout,
     });
-    this._cacheRootFolder = String(
-      opts.cacheFolder || this.getOption('cache-folder', true) || constants.MODULE_CACHE_DIRECTORY,
-    );
+
+    let cacheRootFolder = opts.cacheFolder || this.getOption('cache-folder', true);
+
+    if (!cacheRootFolder) {
+      let preferredCacheFolders = constants.PREFERRED_MODULE_CACHE_DIRECTORIES;
+      const preferredCacheFolder = opts.preferredCacheFolder || this.getOption('preferred-cache-folder', true);
+
+      if (preferredCacheFolder) {
+        preferredCacheFolders = [preferredCacheFolder].concat(preferredCacheFolders);
+      }
+
+      for (let t = 0; t < preferredCacheFolders.length && !cacheRootFolder; ++t) {
+        const tentativeCacheFolder = preferredCacheFolders[t];
+
+        try {
+          await fs.mkdirp(tentativeCacheFolder);
+          // eslint-disable-next-line
+          await fs.access(tentativeCacheFolder, fs.constants.R_OK | fs.constants.W_OK | fs.constants.X_OK);
```

Comment:
On second thought, we may wanna add a comment about this since I'm not sure if "You can't traverse folders if they are not executable." is common knowledge. What do you think?